<html>
<head>
  <title>OMIM search</title>
  <link rel="shortcut icon" href="/omim-search/favicon.ico">

  <!-- DataTables download builder: https://datatables.net/download/  - selected: DataTables, jQuery3, DataTables, Buttons: Column Visibility, HTML5 export, Print view, ColReorder, Responsive, Scroller -->
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" />
  <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/v/se/jq-3.3.1/jszip-2.5.0/dt-1.10.20/b-1.6.1/b-colvis-1.6.1/b-flash-1.6.1/b-html5-1.6.1/b-print-1.6.1/datatables.min.css"/>

  <style>
    .dt-buttons {
      margin-right: 15px !important;
      display: inline-block;
    }

    .dataTables_length {
      margin-right: 10px;
      display: inline-block;
    }

    .dataTables_paginate {
      margin-right: 210px;
      margin-top: 10px;
      margin-bottom: 10px;
      display: inline-block;
    }

    #data-table_wrapper {
      padding: 0px !important;
    }

    #data-table {
      margin-bottom: 20px;
    }

    .no-margin {
      margin: 0px !important;
    }

    .google-visualization-tooltip {
      margin-top: 100px;
      margin-right: 100px;
      z-index:+1;
    }

    .dataTables_empty {
      text-align: left !important;
    }
  </style>

  <!-- "all genes", "all OMIM genes", "all OMIM genes with phenotype", "all OMIM recessive genes" -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.36/pdfmake.min.js"></script>
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.36/vfs_fonts.js"></script>
  <script type="text/javascript" src="https://cdn.datatables.net/v/se/jq-3.3.1/jszip-2.5.0/dt-1.10.20/b-1.6.1/b-colvis-1.6.1/b-flash-1.6.1/b-html5-1.6.1/b-print-1.6.1/datatables.min.js"></script>
  <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery-timeago/1.6.7/jquery.timeago.min.js"></script>
  <script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
  <script>
    window.COLUMN_IDX = {
      MIM_NUMBER: 0,
      PHENOTYPE_MIM_NUMBER: 1,
      PHENOTYPIC_SERIES_NUMBER: 2,
      PHENOTYPE_INHERITANCE: 3,

      LOCUS: 4,
      LOCUS_HG19: 5,

      CYTO: 6,
      GENE_SYMBOLS: 7,
      GENE_ID: 8,
      GENE_DESCRIPTION: 9,
      PHENOTYPE_DESCRIPTION: 10,
      DATE_CREATED: 11,
      DATE_UPDATED: 12,
      MOUSE_GENE_ID: 13,

      TEXT: 14,
      COMMENTS: 15,

      XSTART: 16,
      XEND: 17,
      XSTART_HG19: 18,
      XEND_HG19: 19,
      PHENOTYPE_MAP_METHOD: 20,
    }

    CHROMOSOMES = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y', 'M']

    OMIM_PHENOTYPE_MAP = {
      '1': 'The disorder is placed on the map based on its association with a gene, but the underlying defect is not known.',
      '2': 'The disorder has been placed on the map by linkage; no mutation has been found.',
      '3': 'The molecular basis for the disorder is known; a mutation has been found in the gene.',
      '4': 'A contiguous gene deletion or duplication syndrome, multiple genes are deleted or duplicated causing the phenotype.',
      '': ''
    }


    function getXpos(chrom, pos) {
      /* Compute single number representing this chromosome and position.
          chrom (string): examples '1', 'Y', 'M'
          pos (integer): genomic position on chromosome
      */

      if (chrom.startsWith('chr')) {
        chrom = chrom.replace('chr', '')
      }
      if (chrom.startsWith('M')) {
        chrom = 'M'
      }

      if (pos < 1 || pos > 3e8) {
        console.error('Invalid pos arg: ' + pos)
      }

      return (1 + CHROMOSOMES.indexOf(chrom))*1e9 + pos
    }

    function formatRegionSize(numBases) {
      if (numBases >= 1_000_000) {
        return (numBases/1_000_000.0).toFixed(1) + ' Mb'
      } else if (numBases >= 1_000) {
        return (numBases/1_000.0).toFixed(1) + ' kb'
      } else {
        return numBases + ' bp'
      }
    }

    function runSearch() {
      var searchBoxText = ($('#search-box').val() || '').trim()
      //check for one or more genomic intervals
      window.genomicIntervals = []

      var reg = /(chr)?([1-9XYM][0-9Tt]?)[-:\s]+([0-9]+)[-\s]+([0-9]+)/g
      var match
      var rawIntervalStrings = []
      var parsedIntervalStrings = []
      while((match = reg.exec(searchBoxText)) !== null) {
        var chrom = match[2]
        var xStart = getXpos(chrom, parseInt(match[3]))
        var xEnd = getXpos(chrom, parseInt(match[4]))

        window.genomicIntervals.push([xStart, xEnd])
        parsedIntervalStrings.push(chrom + ':' + match[3] + '-' + match[4] + '  ('  + formatRegionSize(xEnd - xStart) + ')')
        rawIntervalStrings.push(match[0])
      }

      if (rawIntervalStrings.length > 0) {
        $.each(rawIntervalStrings, function(_, s) {
          searchBoxText = searchBoxText.replace(s, '')
        })
        searchBoxText = searchBoxText.replace(',', ' ').trim()
      }

      console.log('searching for:', searchBoxText)
      dataTable.search(searchBoxText)

      window.resultsInfoText = ''
      if (searchBoxText || window.genomicIntervals.length > 0) {
        window.resultsInfoText += ' for '
      }

      if (searchBoxText) {
        // do a regular DataTables search
        window.resultsInfoText += ' <b>keyword(s)</b>: ' + searchBoxText
      }

      if (searchBoxText && window.genomicIntervals.length > 0) {
        window.resultsInfoText += ' and '
      }

      if (window.genomicIntervals.length > 0) {
        window.resultsInfoText += ' <b>region(s)</b>: ' + (parsedIntervalStrings.length <= 3 ? parsedIntervalStrings.join(', ') : (parsedIntervalStrings.length + ' genomic regions') )
      }


      dataTable.draw()

      if (resultsInfoText) {
        $('#data-table_info').append(resultsInfoText)
      }

      drawChart()
    }

    function drawChart() {
      if (!google || !google.visualization || !google.visualization.arrayToDataTable) {
        return
      }


      var HISTOGRAM_DATA_SIZE_THRESHOLD = 500;  // GoogleCharts appears to be too slow and resource-intensive for histograms larger than this

      var data = window.dataTable.rows({filter: 'applied'}).data()

      //create the dataArray
      var dataArray
      if (data.length > HISTOGRAM_DATA_SIZE_THRESHOLD) {
        var counters = {}
        data.each(function(row) {
          var yearCreated = row[COLUMN_IDX.DATE_CREATED].split('-')[0]
          if (yearCreated) {
            if (!counters[yearCreated]) {
              counters[yearCreated] = { created: 0, updated: 0 }
            }
            counters[yearCreated]['created'] += 1
          }
          var yearUpdated = row[COLUMN_IDX.DATE_UPDATED].split('-')[0]
          if (yearUpdated) {
            if (!counters[yearUpdated]) {
              counters[yearUpdated] = { created: 0, updated: 0 }
            }
            counters[yearUpdated]['updated'] += 1
          }
        });

        dataArray = [[ 'Year', 'Year Created', 'Year Updated' ]];
        $.each(Object.keys(counters), function(_, year) {
          dataArray.push([ year, counters[year].created || 0, counters[year].updated || 0 ]);
        })
      } else {
        dataArray = [[ 'Label', 'Year Created', 'Year Updated' ]];
        data.each(function(row) {
          var label = data.length < 500? 'MIM #' + row[COLUMN_IDX.MIM_NUMBER] + ': ' + row[COLUMN_IDX.GENE_SYMBOLS] + ' => ' + row[COLUMN_IDX.PHENOTYPE_DESCRIPTION] : ''
          var yearCreated = parseInt(row[COLUMN_IDX.DATE_CREATED].split('-')[0])
          var yearUpdated = parseInt(row[COLUMN_IDX.DATE_UPDATED].split('-')[0])
          dataArray.push([ label, yearCreated, yearUpdated ]);
        })

      }

      var options = {
        async: true,
        enableInteractivity: true, //data.length <= HISTOGRAM_DATA_SIZE_THRESHOLD,
        hAxis: {
          format: '0000',
          slantedText: true,
          slantedTextAngle: 50,
        },
        //histogram: {
        //  bucketSize: 2,
        //},
        vAxis: {
          format: 'short',
          //minValue: 0,
        },
        tooltip: {
          trigger: data.length > HISTOGRAM_DATA_SIZE_THRESHOLD? 'none': 'focus',
          isHtml: true,
          ignoreBounds: true,
        },
        chartArea: {
          top: 5,
          left: 50,
          backgroundColor: {
            stroke: 'gray',
            strokeWidth: 1,
          },
          width: '90%',
          height: '200px',
        },
        legend: {
          position: 'bottom',
          alignment: 'start',
        },
      };

      if (data.length <= HISTOGRAM_DATA_SIZE_THRESHOLD) {
        options.hAxis.viewWindow = { max: new Date().getFullYear() + 3 }
      }

      var ChartType = data.length > HISTOGRAM_DATA_SIZE_THRESHOLD ? google.visualization.ColumnChart : google.visualization.Histogram

      var data = google.visualization.arrayToDataTable(dataArray);
      var chart = new ChartType(document.getElementById('chart-div'));
      chart.draw(data, options);
    }


    $(document).ready(
      function () {
        $('.ui.radio.checkbox').checkbox()

        google.charts.load("current", {packages:["corechart"]});
        google.charts.setOnLoadCallback(drawChart);

        window.dataTable = $('#data-table').DataTable({
          pageLength: 20,
          lengthMenu: [ 10, 20, 50, 100, 200, 500, 1000, 2000 ],
          ajax: "/omim-search/omim.json",
          dom: 'lrpBitlrpB',
          order: [[ COLUMN_IDX.DATE_CREATED, 'desc' ]],
          responsive: false,
          saveState: true,
          stateDuration: 0,
          columnDefs: [
            {
              targets: [ COLUMN_IDX.MIM_NUMBER, COLUMN_IDX.PHENOTYPE_MIM_NUMBER ],
              render: function ( data, type, row, meta ) {
                if(type === 'display'){
                  var mimNumber = data
                  data = '<a title="' + (OMIM_PHENOTYPE_MAP[row[COLUMN_IDX.PHENOTYPE_MAP_METHOD]] || '') + '" href="http://www.omim.org/entry/' + mimNumber + '" target="_blank">' + mimNumber + '</a>'
                }

                return data;
              }
            },
            {
              targets: [ COLUMN_IDX.PHENOTYPIC_SERIES_NUMBER ],
              render: function ( data, type, row, meta ) {
                if(type === 'display'){
                  var psNumber = data
                  data = '<a href="https://www.omim.org/phenotypicSeries/?phenotypicSeriesNumber=' + psNumber + '" target="_blank">' + psNumber + '</a>'
                }

                return data
              }
            },
            {
              targets: [ COLUMN_IDX.LOCUS ],
              render: function ( data, type, row, meta ) {
                if(type === 'display'){
                  var hg38Locus = data
                  data = '<a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=' + encodeURI(hg38Locus) + '&hgsid=784505957_UfSDk5lLjm3ATXWGjNsQAHd282d0" target="_blank">' + hg38Locus + '</a>'
                }

                return data
              }
            },
            {
              targets: [ COLUMN_IDX.LOCUS_HG19 ],
              render: function ( data, type, row, meta ) {
                if(type === 'display'){
                  var hg19Locus = data
                  if (hg19Locus.indexOf('failed') < 0) {
                    data = '<a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=' + encodeURI(hg19Locus) + '&hgsid=784505957_UfSDk5lLjm3ATXWGjNsQAHd282d0" target="_blank">' + hg19Locus + '</a>'
                  }
                }

                return data
              }
            },
            {
              targets: [ COLUMN_IDX.GENE_ID ],
              render: function ( data, type, row, meta ) {
                if(type === 'display'){
                  var geneId = data
                  data = '<a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=' + geneId + '" target="_blank">' + geneId + '</a>'
                }

                return data
              }
            },
          ],
          buttons: ['copy', 'csv', 'excel', 'pdf', 'print'],
          bAutoWidth: false,
          deferRender: true,
          initComplete: function(settings, json) {
            if(json.date_downloaded_from_omim) {
              $('#omim-data-timestamp').text($.timeago(json.date_downloaded_from_omim))
            }
            drawChart()
          }
        })

        /*
        $("#search-box").keyup(function(e) {
          if (e.keyCode !== 13) {
            return
          }
          runSearch()
        })
        */

        $.fn.dataTableExt.afnFiltering.push(
          function( oSettings, aData, iDataIndex ) {
            if (window.genomicIntervals && window.genomicIntervals.length > 0) {
              var hgRadioValue = $("input[name='hg']:checked").val()
              var isInsideInterval = false
              for (var i = 0; i < window.genomicIntervals.length; i += 1) {
                var interval = window.genomicIntervals[i]
                if (hgRadioValue === 'hg38' && (parseInt(aData[COLUMN_IDX.XSTART]) >= interval[0]) && (parseInt(aData[COLUMN_IDX.XEND]) <= interval[1])) {
                  isInsideInterval = true
                  break
                }
                if (hgRadioValue === 'hg19' && (parseInt(aData[COLUMN_IDX.XSTART_HG19]) >= interval[0]) && (parseInt(aData[COLUMN_IDX.XEND_HG19]) <= interval[1])) {
                  isInsideInterval = true
                  break
                }
              }
              if (!isInsideInterval) {
                return false
              }
            }

            var showRadioValue = $("input[name='show-only']:checked").val()
            if ( showRadioValue === "genes" || showRadioValue === "genes-with-phenotype") {
              var rowHasGene = aData[COLUMN_IDX.GENE_SYMBOLS].trim() || aData[COLUMN_IDX.GENE_ID].trim()
              var rowHasPhenotype = aData[COLUMN_IDX.PHENOTYPE_MIM_NUMBER].trim()
              var showRadioValue = $("input[name='show-only']:checked").val()
              if ( showRadioValue === "genes" && !rowHasGene )
              {
                return false
              }
              else if ( showRadioValue === "genes-with-phenotype" && (!rowHasGene || !rowHasPhenotype) )
              {
                return false
              }
            }

            var inheritance = $("input[name='inheritance']:checked").val()
            if (inheritance !== "any" && aData[COLUMN_IDX.PHENOTYPE_INHERITANCE].toLowerCase().indexOf(inheritance) === -1) {
               return false
            }

            return true
          }
        )

        $.each([ COLUMN_IDX.XSTART, COLUMN_IDX.XEND, COLUMN_IDX.XSTART_HG19, COLUMN_IDX.XEND_HG19, COLUMN_IDX.PHENOTYPE_MAP_METHOD ], function(_, i) {
          dataTable.column( i ).visible( false )
        })

        $("#search-button").click( function() {
          runSearch()
        })
      })

  </script>

  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-154808357-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-154808357-1');
  </script>
</head>
<body style="margin:20px 20px; overflow: auto">

  <div class="ui stackable grid">
    <div class="row">

      <div class="seven wide column">
        <div class="ui form" style="width:100%">
          <div class="field no-margin" style="padding-bottom: 15px;">
            <textarea rows="5" style="width: 100%" id="search-box" placeholder="Enter keywords (eg. dystrophy), phrases (eg. 'myotonic dystrophy'), and regions separated by commas or spaces (eg. chr1:12345-54321, chr2   56789    98765)"></textarea><br />
          </div>
          <div class="ui stackable grid">
            <div class="row" style="margin-bottom: 20px">
              <div class="thirteen wide column">
                <div class="inline fields no-margin">
                  <div class="field" style="padding-right:20px; padding-bottom:10px;">
                    <b>Loci:</b>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="hg" value="hg19" /><label>hg19</label>
                    </div>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="hg" value="hg38" checked /><label>hg38</label>
                    </div>
                  </div>
                </div>
                <div class="inline fields no-margin">
                  <div class="field" style="padding-right:30px; padding-bottom:10px">
                    <b>Show:</b>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="show-only" value="all" checked /><label>All search hits</label>
                    </div>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="show-only" value="genes" /><label>OMIM genes</label>
                    </div>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="show-only" value="genes-with-phenotype"><label>OMIM genes with phenotype</label>
                    </div>
                  </div>
                </div>
                <div class="inline fields no-margin">
                  <div class="field" style="padding-right:20px; padding-bottom:10px">
                    <b>Inheritance:</b>
                    <div class="ui radio checkbox no-margin" style="padding-left:15px">
                      <input type="radio" name="inheritance" value="any" checked /><label>Any</label>
                    </div>
                    <div class="ui radio checkbox no-margin" style="padding-left:10px">
                      <input type="radio" name="inheritance" value="recessive" /><label>Recessive</label>
                    </div>
                    <div class="ui radio checkbox no-margin" style="padding-left:10px">
                      <input type="radio" name="inheritance" value="dominant" /><label>Dominant</label>
                    </div>
                  </div>
                </div>
              </div>
              <div class="column">
                <div style="padding: 15px 0px;">
                  <button id="search-button" class="ui primary button" type="submit" style="white-space: nowrap"> <i class="search icon"></i> Search</button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="four wide column">
        <div style="margin-bottom: 10px"><b>OMIM data last downloaded </b>
          <span id="omim-data-timestamp"> ..</span>.
        </div>
        <div style="margin-bottom: 10px"><b>DISCLAIMER:</b> Intended for research use only.</div>
        <div style="margin-bottom: 10px">Issues or feature requests can be sent via
          <a href="https://github.com/macarthur-lab/omim-search/issues" target="_blank">github</a>.
        </div>
      </div>
      <div class="five wide column">
        <div id="chart-div" style="width: 400px; height: 250px;"></div>
      </div>
    </div>
  </div>

  <table id="data-table" class="ui celled table">
    <thead>
    <tr>
      <!-- 0 --><th style="min-width:50px">MIM #</th>
      <!-- 1 --><th style="min-width:50px">Phenotype MIM #</th>
      <!-- 2 --><th style="min-width:50px">Phenotypic Series #</th>
      <!-- 3 --><th style="min-width:150px">Inheritance</th>

      <!-- 4 --><th style="min-width:100px">hg38 Locus</th>
      <!-- 5 --><th style="min-width:100px">hg19 Locus</th>

      <!-- 6 --><th style="min-width:50px">Cyto</th>
      <!-- 7 --><th style="min-width:100px">Gene</th>
      <!-- 8 --><th style="min-width:50px">Gene Id</th>
      <!-- 9 --><th style="min-width:200px">Gene Description</th>
      <!-- 10 --><th style="min-width:200px">Phenotype Description</th>
      <!-- 11 --><th style="min-width:70px">Date Created</th>
      <!-- 12 --><th style="min-width:70px">Date Updated</th>
      <!-- 13 --><th style="min-width:100px">Mouse Gene Id</th>
      <!-- 14 --><th style="min-width:1000px">Text</th>
      <!-- 15 --><th style="min-width:170px">Comments</th>

      <!-- 16 --><th></th>
      <!-- 17 --><th></th>
      <!-- 18 --><th></th>
      <!-- 19 --><th></th>
      <!-- 20 --><th></th>

    </tr>
    </thead>
  </table>
  <br />
  <br />
</body>
</html>
