<html><head><meta charset="utf-8"><title>OMIM search</title><link rel="shortcut icon" href="/favicon.4f230e9a.ico"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css"><link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/v/se/jq-3.3.1/jszip-2.5.0/dt-1.10.20/b-1.6.1/b-colvis-1.6.1/b-flash-1.6.1/b-html5-1.6.1/b-print-1.6.1/datatables.min.css"><style>.dt-buttons{margin-right:15px!important;display:inline-block}.dataTables_length{margin-right:10px;display:inline-block}.dataTables_paginate{margin-right:210px;margin-top:10px;margin-bottom:10px;display:inline-block}#data-table_wrapper{padding:0!important}#data-table{margin-bottom:20px}.no-margin{margin:0!important}.google-visualization-tooltip{margin-top:100px;margin-right:100px;z-index:1}.dataTables_empty{text-align:left!important}</style><script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.36/pdfmake.min.js"></script><script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.36/vfs_fonts.js"></script><script type="text/javascript" src="https://cdn.datatables.net/v/se/jq-3.3.1/jszip-2.5.0/dt-1.10.20/b-1.6.1/b-colvis-1.6.1/b-flash-1.6.1/b-html5-1.6.1/b-print-1.6.1/datatables.min.js"></script><script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script><script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery-timeago/1.6.7/jquery.timeago.min.js"></script><script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script><script>function e(e,r){return n(e)||a(e,r)||t()}function t(){throw new TypeError("Invalid attempt to destructure non-iterable instance")}function a(e,t){if(Symbol.iterator in Object(e)||"[object Arguments]"===Object.prototype.toString.call(e)){var a=[],n=!0,r=!1,o=void 0;try{for(var i,l=e[Symbol.iterator]();!(n=(i=l.next()).done)&&(a.push(i.value),!t||a.length!==t);n=!0);}catch(s){r=!0,o=s}finally{try{n||null==l.return||l.return()}finally{if(r)throw o}}return a}}function n(e){if(Array.isArray(e))return e}function r(e,t){return e.startsWith("chr")&&(e=e.replace("chr","")),e.startsWith("M")&&(e="M"),(t<1||t>3e8)&&console.error("Invalid pos arg: "+t),1e9*(1+CHROMOSOMES.indexOf(e))+t}function o(e){return e>=1e6?(e/1e6).toFixed(1)+" Mb":e>=1e3?(e/1e3).toFixed(1)+" kb":e+" bp"}function i(){return window.location.hash.substr(1).split("&").reduce(function(e,t){var a=t.split("=");return e[a[0]]=decodeURIComponent(a[1]),e},{})}function l(t){var a=Object.entries(t).map(function(t){var a=e(t,2),n=a[0],r=a[1];return n+"="+encodeURIComponent(r)}).join("&");console.log("Setting hash to",a),window.location.hash="#"+a}function s(){if(window.localStorage){var e=localStorage.getItem("hg");e&&$("input[value='"+e+"']").prop("checked",!0)}}function d(){window.inputValues={hg:$("input[name='hg']:checked").val(),mustHaveGeneId:$("#must-have-gene-id").prop("checked"),mustHavePhenotype:$("#must-have-phenotype").prop("checked"),inheritance:$("input[name='inheritance']:checked").val()},window.localStorage&&localStorage.setItem("hg",inputValues.hg)}function c(e){var t=($("#search-box").val()||"").trim(),a=$("#search-column option:selected").val();e||$("#data-table-container").attr("style","display: block !important"),window.genomicIntervals=[];for(var n,i=/(chr)?([1-9XYM][0-9Tt]?)[-:\s]+([0-9,]+)[-\s]+([0-9,]+)/g,l=[],s=[];null!==(n=i.exec(t));){var c=n[2],p=n[3].replace(/,/g,""),g=n[4].replace(/,/g,""),h=r(c,parseInt(p)),E=r(c,parseInt(g));window.genomicIntervals.push([h,E]),s.push(c+":"+p+"-"+g+"  ("+o(E-h)+")"),l.push(n[0])}if(l.length>0&&($.each(l,function(e,a){t=t.replace(a,"")}),t=t.replace(","," ").trim()),console.log("searching for",'"'+t+'"',"in",a,"column(s)"),dataTable.search(""),dataTable.columns(Object.values(COLUMN_IDX)).search(""),"all"===a)dataTable.search(t);else{var I=COLUMN_IDX[a];console.log("Searching ",a," idx",I),dataTable.column(I).search(t)}window.resultsInfoText="",(t||window.genomicIntervals.length>0)&&(window.resultsInfoText+=" for "),t&&(window.resultsInfoText+=" <b>keyword(s)</b>: "+t),t&&window.genomicIntervals.length>0&&(window.resultsInfoText+=" and "),window.genomicIntervals.length>0&&(window.resultsInfoText+=" <b>region(s)</b>: "+(s.length<=3?s.join(", "):s.length+" genomic regions")),d(),dataTable.draw(),resultsInfoText&&$("#data-table_info").append(resultsInfoText),u()}function u(){if(google&&google.visualization&&google.visualization.arrayToDataTable){var e;if((r=window.dataTable.rows({filter:"applied"}).data()).length>500){var t={};r.each(function(e){var a=e[COLUMN_IDX.DATE_CREATED].split("-")[0];a&&(t[a]||(t[a]={created:0,updated:0}),t[a].created+=1);var n=e[COLUMN_IDX.DATE_UPDATED].split("-")[0];n&&(t[n]||(t[n]={created:0,updated:0}),t[n].updated+=1)}),e=[["Year","Year Created","Year Updated"]],$.each(Object.keys(t),function(a,n){e.push([n,t[n].created||0,t[n].updated||0])})}else e=[["Label","Year Created","Year Updated"]],r.each(function(t){var a=r.length<500?"MIM #"+t[COLUMN_IDX.MIM_NUMBER]+": "+t[COLUMN_IDX.GENE_SYMBOLS]+" => "+t[COLUMN_IDX.PHENOTYPE_DESCRIPTION]:"",n=parseInt(t[COLUMN_IDX.DATE_CREATED].split("-")[0]),o=parseInt(t[COLUMN_IDX.DATE_UPDATED].split("-")[0]);e.push([a,n,o])});var a={async:!0,enableInteractivity:!0,hAxis:{format:"0000",slantedText:!0,slantedTextAngle:50},vAxis:{title:"OMIM Records",format:"short",titleTextStyle:{fontSize:12,italic:!1,bold:!1}},tooltip:{isHtml:!0,ignoreBounds:!0},chartArea:{top:5,left:50,backgroundColor:{stroke:"gray",strokeWidth:1},width:"90%",height:"200px"},legend:{position:"bottom",alignment:"start"}};r.length<=500&&(a.hAxis.viewWindow={max:(new Date).getFullYear()+3});var n=r.length>500?google.visualization.ColumnChart:google.visualization.Histogram,r=google.visualization.arrayToDataTable(e);new n(document.getElementById("chart-div")).draw(r,a)}}window.COLUMN_IDX={MIM_NUMBER:0,PHENOTYPE_MIM_NUMBER:1,PHENOTYPIC_SERIES_NUMBER:2,PHENOTYPE_INHERITANCE:3,LOCUS_SIZE:4,LOCUS:5,LOCUS_HG19:6,CYTO:7,GENE_SYMBOLS:8,GENE_ID:9,GENE_DESCRIPTION:10,PHENOTYPE_DESCRIPTION:11,DATE_CREATED:12,DATE_UPDATED:13,MOUSE_GENE_ID:14,TEXT:15,COMMENTS:16,XSTART:17,XEND:18,XSTART_HG19:19,XEND_HG19:20,PHENOTYPE_MAP_METHOD:21},CHROMOSOMES=["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X","Y","M"],OMIM_PHENOTYPE_MAP_LABEL={1:"association",2:"linkage",3:"variant found",4:"multiple genes deleted or duplicated","":""},OMIM_PHENOTYPE_MAP_DESCRIPTION={1:"The disorder is placed on the map based on its association with a gene, but the underlying defect is not known.",2:"The disorder has been placed on the map by linkage; no mutation has been found.",3:"The molecular basis for the disorder is known; a mutation has been found in the gene.",4:"A contiguous gene deletion or duplication syndrome, multiple genes are deleted or duplicated causing the phenotype.","":""},$(document).ready(function(){$(".ui.radio.checkbox").checkbox(),google.charts.load("current",{packages:["corechart"]}),google.charts.setOnLoadCallback(u),s(),d(),window.dataTable=$("#data-table").DataTable({stateSave:!0,pageLength:20,lengthMenu:[10,20,50,100,200,500,1e3,2e3],ajax:{url:"https://broadinstitute.github.io/omim-search-p/d",dataSrc:function(e){return JSON.parse(atob(e[0])).data}},dom:"lrpBitlrpB",order:[[COLUMN_IDX.DATE_CREATED,"desc"]],responsive:!1,saveState:!0,stateDuration:0,columnDefs:[{targets:[COLUMN_IDX.MIM_NUMBER,COLUMN_IDX.PHENOTYPE_MIM_NUMBER],render:function(e,t,a,n){if("display"===t){e='<a href="http://www.omim.org/entry/'+e+'" target="_blank">'+e+"</a>"}return e}},{targets:[COLUMN_IDX.PHENOTYPIC_SERIES_NUMBER],render:function(e,t,a,n){if("display"===t){e='<a href="https://www.omim.org/phenotypicSeries/?phenotypicSeriesNumber='+e+'" target="_blank">'+e+"</a>"}return e}},{targets:[COLUMN_IDX.LOCUS],render:function(e,t,a,n){if("display"===t){var r=e;e='<a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position='+encodeURI(r)+'&hgsid=784505957_UfSDk5lLjm3ATXWGjNsQAHd282d0" target="_blank">'+r+"</a>"}return e}},{targets:[COLUMN_IDX.LOCUS_HG19],render:function(e,t,a,n){if("display"===t){var r=e;r.indexOf("failed")<0&&(e='<a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position='+encodeURI(r)+'&hgsid=784505957_UfSDk5lLjm3ATXWGjNsQAHd282d0" target="_blank">'+r+"</a>")}return e}},{targets:[COLUMN_IDX.LOCUS_SIZE],render:function(e,t,a,n){return"display"===t&&(e=parseInt(e).toLocaleString()),e}},{targets:[COLUMN_IDX.GENE_SYMBOLS,COLUMN_IDX.GENE_ID],render:function(e,t,a,n){return"display"===t&&(e='<a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene='+(e?e.split(",")[0].trim():"")+'" target="_blank">'+e+"</a>"),e}},{targets:[COLUMN_IDX.PHENOTYPE_MAP_METHOD],render:function(e,t,a,n){return"display"===t&&e&&(e=OMIM_PHENOTYPE_MAP_LABEL[e]),e}}],buttons:["copy","csv","excel","pdf","print"],bAutoWidth:!1,deferRender:!0,initComplete:function(e,t){(t=JSON.parse(atob(t[0]))).date_downloaded_from_omim&&$("#omim-data-timestamp").text($.timeago(t.date_downloaded_from_omim)),c(!0)}}),$.fn.dataTableExt.afnFiltering.push(function(e,t,a){if(window.genomicIntervals&&window.genomicIntervals.length>0){for(var n=!1,r=0;r<window.genomicIntervals.length;r+=1){var o=window.genomicIntervals[r];if("hg38"===inputValues.hg&&parseInt(t[COLUMN_IDX.XSTART])<=o[1]&&parseInt(t[COLUMN_IDX.XEND])>=o[0]){n=!0;break}if("hg19"===inputValues.hg&&parseInt(t[COLUMN_IDX.XSTART_HG19])<=o[1]&&parseInt(t[COLUMN_IDX.XEND_HG19])>=o[0]){n=!0;break}}if(!n)return!1}return!(inputValues.mustHaveGeneId&&!t[COLUMN_IDX.GENE_ID].trim())&&(!(inputValues.mustHavePhenotype&&!t[COLUMN_IDX.PHENOTYPE_MIM_NUMBER].trim())&&("any"===inputValues.inheritance||-1!==t[COLUMN_IDX.PHENOTYPE_INHERITANCE].toLowerCase().indexOf(inputValues.inheritance)))}),$.each([COLUMN_IDX.XSTART,COLUMN_IDX.XEND,COLUMN_IDX.XSTART_HG19,COLUMN_IDX.XEND_HG19],function(e,t){dataTable.column(t).visible(!1)}),$("#search-button").click(function(){c()}),$(".question").popup()});</script><script async src="https://www.googletagmanager.com/gtag/js?id=UA-154808357-1"></script><script>function a(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],a("js",new Date),a("config","UA-154808357-1");</script></head><body style="margin:20px 20px;overflow:auto"> <div class="ui stackable grid"> <div class="row"> <div class="seven wide column"> <div class="ui form" style="width:100%"> <div class="field no-margin" style="padding-bottom:15px;"> <textarea rows="5" style="width:100%" id="search-box" placeholder="Enter keywords (eg. dystrophy), phrases in double quotes (eg. &quot;myotonic dystrophy&quot;), and/or regions separated by commas or spaces (each region must be in the format chr1:12345-54321 or chr1   12345    54321). The search will return OMIM records that overlap these region(s) and include the given keywords and phrases."></textarea><br> </div> <div class="ui stackable grid"> <div class="row" style="margin-bottom:20px"> <div class="thirteen wide column"> <div class="inline fields no-margin"> <div class="field" style="padding-right:30px;padding-bottom:12px;"> <span style="display:inline-block;white-space:nowrap;padding-bottom:10px;"><b>Search in:</b></span> <select id="search-column" class="ui dropdown" style="height:auto!important;padding:3px;"> <option value="all">All Columns</option> <option value="MIM_NUMBER">MIM #</option> <option value="PHENOTYPE_MIM_NUMBER">Phenotype MIM #</option> <option value="PHENOTYPIC_SERIES_NUMBER">Phenotypic Series #</option> <option value="PHENOTYPE_INHERITANCE">Inheritance</option> <option value="CYTO">Cyto</option> <option value="GENE_SYMBOLS">Gene</option> <option value="GENE_ID">Gene Id</option> <option value="GENE_DESCRIPTION">Gene Description</option> <option value="PHENOTYPE_DESCRIPTION">Phenotype Description</option> <option value="DATE_CREATED">Date Created</option> <option value="DATE_UPDATED">Date Updated</option> <option value="MOUSE_GENE_ID">Mouse Gene Id</option> <option value="TEXT">Text</option> <option value="COMMENTS">Comments</option> </select> </div> <div class="field" style="padding-right:10px;padding-bottom:12px"> <span style="display:inline-block;white-space:nowrap;padding-bottom:10px;"><b>Search regions:</b></span> <div class="ui radio checkbox no-margin" style="padding-left:15px;padding-bottom:10px"> <input type="radio" name="hg" value="hg38" checked><label>hg38</label> </div> <div class="ui radio checkbox no-margin" style="padding-left:15px;padding-bottom:10px"> <input type="radio" name="hg" value="hg19"><label>hg19</label> </div> </div> </div> <div class="inline fields no-margin"> <div class="field" style="padding-right:20px;padding-bottom:12px;"> <span style="display:inline-block;white-space:nowrap;padding-bottom:10px;"><b>Results must have:</b></span> <div class="ui checkbox" style="margin-left:15px;margin-bottom:10px"> <input type="checkbox" id="must-have-gene-id" checked> <label for="must-have-gene-id" style="cursor:pointer">Gencode (ENSG) Gene Id</label> </div> <div class="ui checkbox" style="margin-left:15px;margin-bottom:10px"> <input type="checkbox" id="must-have-phenotype"> <label for="must-have-phenotype" style="cursor:pointer">Phenotype MIM #</label> </div> </div> </div>  <div class="inline fields no-margin"> <div class="field" style="padding-right:20px"> <span style="display:inline-block;white-space:nowrap;padding-bottom:10px;"><b>Inheritance:</b></span> <div class="ui radio checkbox no-margin" style="padding-left:15px;padding-bottom:10px"> <input type="radio" name="inheritance" value="any" checked><label>any</label> </div> <div class="ui radio checkbox no-margin" style="padding-left:10px;padding-bottom:10px"> <input type="radio" name="inheritance" value="recessive"><label>recessive</label> </div> <div class="ui radio checkbox no-margin" style="padding-left:10px;padding-bottom:10px"> <input type="radio" name="inheritance" value="dominant"><label>dominant</label> </div> </div> </div> </div> <div class="three wide column"> <div style="padding:0 0;"> <button id="search-button" class="ui primary button" type="submit" style="white-space:nowrap"> <i class="search icon"></i> Search</button> </div> </div> </div> </div> </div> </div> <div class="four wide column"> <div style="margin-bottom:10px"><b>OMIM data last downloaded </b> <span id="omim-data-timestamp"> ..</span>. </div> <div style="margin-bottom:10px"><b>DISCLAIMER:</b> Intended for research use only.</div> <div style="margin-bottom:10px"> <a href="https://github.com/macarthur-lab/omim-search/issues" target="_blank">Issues</a> and <a href="https://github.com/macarthur-lab/omim-search/issues" target="_blank">feature requests</a> can be submitted via <a href="https://github.com/macarthur-lab/omim-search" target="_blank">github</a>. </div> </div> <div class="five wide column"> <div style="text-align:center"> <b>OMIM Records Timeline</b> &nbsp; <i class="question circle outline icon" data-position="bottom center" data-content="This graph shows the number of OMIM records that were created or updated in a given year. Each OMIM record is counted twice - once in the 'Created' bars based on when it was first added to OMIM, and once in the 'Updated' bars based on when it was last updated."></i> </div> <div id="chart-div" style="width:400px;height:250px;"></div> </div> </div> </div> <div id="data-table-container" style="display:none"> <table id="data-table" class="ui celled table"> <thead> <tr> <th style="min-width:50px">MIM #</th> <th style="min-width:50px">Phenotype MIM #</th> <th style="min-width:50px">Phenotypic Series #</th> <th style="min-width:150px">Inheritance</th> <th style="min-width:100px">Locus Size</th> <th style="min-width:100px">hg38 Locus</th> <th style="min-width:100px">hg19 Locus</th> <th style="min-width:50px">Cyto</th> <th style="min-width:100px">Gene</th> <th style="min-width:50px">Gene Id</th> <th style="min-width:200px">Gene Description</th> <th style="min-width:200px">Phenotype Description</th> <th style="min-width:70px">Date Created</th> <th style="min-width:70px">Date Updated</th> <th style="min-width:100px">Mouse Gene Id</th> <th style="min-width:1000px">Text</th> <th style="min-width:170px">Comments</th> <th>xStart</th> <th>xEnd</th> <th>xStart hg19</th> <th>xEnd hg19</th> <th>Phenotype <span style="white-space:nowrap">Map Method</span></th> </tr> </thead> </table> </div> <br> <br>
</body></html>